### PR TITLE
Fix alternative search engine provider feature bugs

### PR DESCRIPTION
**Pull request by simonhong**
First commit addressed https://github.com/brave/brave-browser/issues/1235.
Second commit addressed https://github.com/brave/brave-browser/issues/1393
Third commit is refactoring and cleanup.
Forth commit addressed https://github.com/brave/brave-browser/issues/1394
Fifth commit partially addressed https://github.com/brave/brave-browser/issues/1226 because current PR only uses ddg for tor window.

Close: https://github.com/brave/brave-browser/issues/1235
Close: https://github.com/brave/brave-browser/issues/1393
Close: https://github.com/brave/brave-browser/issues/1394
Issue: https://github.com/brave/brave-browser/issues/1226

Simple high level overview

When profile is created, appropriate `SearchEngineProviderController` is initialized.
`PrivateWindowSearchEngineProviderController` is for private window and `GuestWindowSearchEngineProviderController` is for guest window.
`ProfileCreationMonitor` monitors profile creation and tries to initialize above controllers if needed.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
Open private window and toggle ddg button and reload.
Then check toggle state is remained as before.
`yarn test brave_browser_tests --filter=SearchEngineProviderControllerTest.*`

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

**Comment by simonhong** [comment]
Marked as wip. I'll push all fixes of ddg search provider issues to this PR.

File: browser/search_engine_provider_controller_base.cc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "chrome/browser/search_engines/template_url_service_factory.h"
+#include "components/prefs/pref_service.h"
+#include "components/search_engines/template_url_service.h"
+
+namespace {
+
+TemplateURLData GetAlternativeSearchEngineData() {
+  TemplateURLData private_search_engine_data;
+  private_search_engine_data.SetShortName(base::ASCIIToUTF16("DuckDuckGo"));
+  private_search_engine_data.SetKeyword(base::ASCIIToUTF16("duckduckgo.com"));
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/search_engine_provider_controller_base.cc]: I deleted this function and used our prepopulated lists to get ddg data.
https://github.com/brave/brave-core/pull/529/commits/4f2480d8804e39e39acfbae8533b8128561a2928

**Comment by bbondy** [comment]
master: aa8863d906e3307407f92a7446ffb2d3e64b2448
0.56.x: 4820c6403148943d91e36fb56ca754608aeb24c0
0.55.x: 1bb0d4b2f306aaf1ff58c684f3dcd78078a6b016

**Comment by simonhong** [comment]
@riastradh-brave Partially addressed. With this PR, only ddg is used on tor.

File: browser/guest_window_search_engine_provider_controller.cc [hunk 2]
```diff
@@ -30,11 +30,20 @@ void GuestWindowSearchEngineProviderController::OnTemplateURLServiceChanged() {
   if (ignore_template_url_service_changing_)
     return;
 
+  // Prevent search engine changing from settings page for tor profile.
+  // TODO(simonhong): Revisit when related ux is determined.
+  if (otr_profile_->IsTorProfile()) {
+    base::AutoReset<bool> reset(&ignore_template_url_service_changing_, true);
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/guest_window_search_engine_provider_controller.cc]: Yep, making this optional is next task.

File: browser/search_engine_provider_controller_browsertest.cc [hunk 3]
```diff
@@ -63,3 +74,21 @@ IN_PROC_BROWSER_TEST_F(SearchEngineProviderControllerTest, PrefTest) {
   EXPECT_EQ(incognito_service->GetDefaultSearchProvider()->data().short_name(),
             base::ASCIIToUTF16("test1"));
 }
+
+// This test crashes with below. I don't know how to deal with now.
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/search_engine_provider_controller_browsertest.cc]: Thanks. I'll update this test later!

**Pull request merged**